File: src/main/java/src/SistemaReservaHotel/Hotel.java
package SistemaReservaHotel;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

public class Hotel {
    private List<Habitacion> habitaciones;
    private List<Cliente> clientes;
    private List<Reserva> reservas;

    public Hotel() {
        this.habitaciones = new ArrayList<>();
        this.clientes = new ArrayList<>();
        this.reservas = new ArrayList<>();
    }

    public void agregarHabitacion(Habitacion habitacion) {
        habitaciones.add(habitacion);
        System.out.println("Se ha añadido la habitación -> " + habitacion);
    }

    public void registrarCliente(Cliente cliente) {
        clientes.add(cliente);
        System.out.println("Se ha añadido el cliente -> " + cliente);
    }

    public void reservarHabitacion(int habitacionId, int clienteId, LocalDate fechaEntrada, LocalDate fechaSalida) {
        Habitacion habitacion = null;
        Cliente cliente = null;

        for (Habitacion habitacionBuscar : habitaciones) {
            if (habitacionBuscar.getId() == habitacionId) {
                habitacion = habitacionBuscar;
                break;
            }
        }

        for (Cliente clienteBuscar : clientes) {
            if (clienteBuscar.getId() == clienteId) {
                cliente = clienteBuscar;
                break;
            }
        }

        if (habitacion == null) {
            System.out.println("La habitación con ID " + habitacionId + " no existe.");
            return;
        }

        if (cliente == null) {
            System.out.println("El cliente con ID " + clienteId + " no está registrado.");
            return;
        }

        if (!habitacion.isDisponible()) {
            System.out.println("La habitación '" + habitacion.getId() + "' ya está reservada.");
            return;
        }

        Reserva reserva = new Reserva(habitacion, cliente, fechaEntrada, fechaSalida);
        reservas.add(reserva);
        habitacion.setDisponible(false);
        System.out.println("Se ha realizado la reserva -> " + reserva);
    }

    public void cancelarReserva(int habitacionId, int clienteId) {
        Reserva reservaCancelar = null;

        for (Reserva reservaBuscar : reservas) {
            if (reservaBuscar.getHabitacion().getId() == habitacionId && reservaBuscar.getCliente().getId() == clienteId) {
                reservaCancelar = reservaBuscar;
                break;
            }
        }

        if (reservaCancelar != null) {
            reservas.remove(reservaCancelar);
            reservaCancelar.getHabitacion().setDisponible(true);
            System.out.println("Se ha cancelado la reserva -> " + reservaCancelar);
        } else {
            System.out.println("No se encontró una reserva para la habitación " + habitacionId + " y cliente " + clienteId + ".");
        }
    }

    public void mostrarReservasActivas() {
        System.out.println("Reservas activas:");
        for (Reserva reserva : reservas) {
            System.out.println(reserva);
        }
    }

    private Habitacion buscarHabitacionPorId(int habitacionId) {
        for (Habitacion habitacion : habitaciones) {
            if (habitacion.getId() == habitacionId) {
                return habitacion;
            }
        }
        return null;
    }

    private Cliente buscarClientePorId(int clienteId) {
        for (Cliente cliente : clientes) {
            if (cliente.getId() == clienteId) {
                return cliente;
            }
        }
        return null;
    }

    private Reserva buscarReserva(int habitacionId, int clienteId) {
        for (Reserva reserva : reservas) {
            if (reserva.getHabitacion().getId() == habitacionId && reserva.getCliente().getId() == clienteId) {
                return reserva;
            }
        }
        return null;
    }
}

File: src/main/kotlin/SistemaReservaHotel/Hotel.kt
package SistemaReservaHotel

import java.time.LocalDate

class Hotel {
    private val habitaciones: MutableList<Habitacion> = mutableListOf()
    private val clientes: MutableList<Cliente> = mutableListOf()
    private val reservas: MutableList<Reserva> = mutableListOf()

    fun agregarHabitacion(habitacion: Habitacion) {
        habitaciones.add(habitacion)
        println("Se ha añadido la habitación -> $habitacion")
    }

    fun registrarCliente(cliente: Cliente) {
        clientes.add(cliente)
        println("Se ha añadido el cliente -> $cliente")
    }

    fun reservarHabitacion(habitacionId: Int, clienteId: Int, fechaEntrada: LocalDate, fechaSalida: LocalDate) {
        var habitacion: Habitacion? = null
        var cliente: Cliente? = null

        for (habitacionBuscar in habitaciones) {
            if (habitacionBuscar.id == habitacionId) {
                habitacion = habitacionBuscar
                break
            }
        }

        for (clienteBuscar in clientes) {
            if (clienteBuscar.id == clienteId) {
                cliente = clienteBuscar
                break
            }
        }

        if (habitacion == null) {
            println("La habitación con ID $habitacionId no existe.")
            return
        }

        if (cliente == null) {
            println("El cliente con ID $clienteId no está registrado.")
            return
        }

        if (!habitacion.disponible) {
            println("La habitación '" + habitacion.id + "' ya está reservada.")
            return
        }

        val reserva = Reserva(habitacion, cliente, fechaEntrada, fechaSalida)
        reservas.add(reserva)
        habitacion.changeDisponible(false)
        println("Se ha realizado la reserva -> $reserva")
    }

    fun cancelarReserva(habitacionId: Int, clienteId: Int) {
        var reservaCancelar: Reserva? = null

        for (reservaBuscar in reservas) {
            if (reservaBuscar.habitacion.id == habitacionId && reservaBuscar.cliente.id == clienteId) {
                reservaCancelar = reservaBuscar
                break
            }
        }

        if (reservaCancelar != null) {
            reservas.remove(reservaCancelar)
            reservaCancelar.habitacion.changeDisponible(true)
            println("Se ha cancelado la reserva -> $reservaCancelar")
        } else {
            println("No se encontró una reserva para la habitación $habitacionId y cliente $clienteId.")
        }
    }

    fun mostrarReservasActivas() {
        println("Reservas activas:")
        for (reserva in reservas) {
            println(reserva)
        }
    }

    private fun buscarHabitacionPorId(habitacionId: Int): Habitacion? {
        for (habitacion in habitaciones) {
            if (habitacion.id == habitacionId) {
                return habitacion
            }
        }
        return null
    }

    private fun buscarClientePorId(clienteId: Int): Cliente? {
        for (cliente in clientes) {
            if (cliente.id == clienteId) {
                return cliente
            }
        }
        return null
    }

    private fun buscarReserva(habitacionId: Int, clienteId: Int): Reserva? {
        for (reserva in reservas) {
            if (reserva.habitacion.id == habitacionId && reserva.cliente.id == clienteId) {
                return reserva
            }
        }
        return null
    }
}